### ISSUE TITLE
[QA] write to external SFTP storage is slow

### ISSUE THREAD

**Issue by jnweiger** [owncloud/core#38420]
setup: owncloud core 10.6.0

with 
```
apt-get install pure-ftpd
adduser ftpdata
```
 -> Settings -> Storage
  SFTP, Username and password
  Configuration
        localhost       /home/ftpdata

* A user moves a 200MB file from his primary storage to SFTP using the web UI.
System loadavg is 0.01
![image](https://user-images.githubusercontent.com/1108546/108136506-77bc7700-70ba-11eb-8b8f-12a39c83223b.png)
* strace shows an sshd process with
```
read(4, "\2432\10\277\0056]\374\237\216\226\362\226w\261\237\24;jS\n\362\345]\21\t\10\253\226\322\36{"..., 16384) = 96
write(11, "\0\0\0(\3\0\0\0\1\0\0\0\27/home/ftpdata/200MB"..., 44) = 44
read(12, "\0\0\0\rf\0\0\0\1\0\0\0\4\0\0\0\0", 16384) = 17
write(4, "Y\247\216\36\16\370#\364%\331\10\0E\31\272\314R\317\220\277W\204\361.[\261\227z\376!\366g"..., 80) = 80
read(4, "\4S\3\320\377\\>\312F\0217\200>\226?\373CQ+!\330\341\351\376\350\353\22\16\303L\364\271"..., 16384) = 8192
read(4, "^\2605\257!R:@\311\364\331\362\1\353\335\221\335\237U\244\36\201\31\224\310c\244p\361\1gy"..., 16384) = 80
write(11, "\0\0 \31\6\0\0\0\0\0\0\0\4\0\0\0\0\0\0\0\0\4\332 \0\0\0 \0\266\206\346"..., 8221) = 8221
read(12, "\0\0\0\30e\0\0\0\0\0\0\0\0\0\0\0\7Success\0\0\0\0", 16384) = 28
write(4, "\235\300H\273\341\3314\342s\3\35LL\323\35\25R\337$\361\267\200\256m\247C\25\225\335\263\36\376"..., 80) = 80
read(4, "\346\370T\267\3514\314\245\335\206\241p\344\301\273\17\331F6\303\314\260hY\240\16\234\215\270\276:v"..., 16384) = 80
write(11, "\0\0\0\r\4\0\0\0\1\0\0\0\4\0\0\0\0", 17) = 17
read(12, "\0\0\0\30e\0\0\0\1\0\0\0\0\0\0\0\7Success\0\0\0\0", 16384) = 28
write(4, "\205\345\350z\325\362\372\212\230\21\0107\316\245\201\370\243\237\206Wy\332\333%0R\234 \354\270z\331"..., 80) = 80
read(4, "\252,)u\240\342\1776s0\303\264\275\263\240\311\273&g\t0\243_\24P!N\334\22\21(c"..., 16384) = 96
write(11, "\0\0\0(\3\0\0\0\1\0\0\0\27/home/ftpdata/200MB"..., 44) = 44
read(12, "\0\0\0\rf\0\0\0\1\0\0\0\4\0\0\0\0", 16384) = 17
write(4, "\372Eq\267\"\272A\215\35\377\0\21|\206.\316y\2\16\335O\224\3374\217\220z9\304\316\26l"..., 80) = 80
read(4, "0\333e\2l\311\6\232s\373T\272D4\315\326[A\23 \23\275\179\366\31,\330\2068\330\246"..., 16384) = 8192
read(4, "\207\304#\376\10.\37n\252\353v\22\2610V\233\36\272\320w\354~\213(G\377\213i}a\312\252"..., 16384) = 80
write(11, "\0\0 \31\6\0\0\0\0\0\0\0\4\0\0\0\0\0\0\0\0\4\332@\0\0\0 \0B\266\333"..., 8221) = 8221
read(12, "\0\0\0\30e\0\0\0\0\0\0\0\0\0\0\0\7Success\0\0\0\0", 16384) = 28
write(4, "F2\312l\354\177\323\303r\324\0375\2635\316(\24\0377\337\4\326\356\214\371\371)\222X<\vm"..., 80) = 80
read(4, "\270\4\324\252\327\375\310\341\r&\363@D\376\374\5\324wN6\321\336Z(\202\214I\245-\261\21X"..., 16384) = 80
write(11, "\0\0\0\r\4\0\0\0\1\0\0\0\4\0\0\0\0", 17) = 17
read(12, "\0\0\0\30e\0\0\0\1\0\0\0\0\0\0\0\7Success\0\0\0\0", 16384) = 28
write(4, "~\342\301l4\225U?\222%&\240\231\1]\276'(g\266 \22X\246\36\216ZX\342Y\345\0"..., 80) = 80
read(4, "G\363v\222\354\321#\327`S\324d\354\360\252n5/\261\2202\321]P\2S\377\341\26\220\277\361"..., 16384) = 96
write(11, "\0\0\0(\3\0\0\0\1\0\0\0\27/home/ftpdata/200MB"..., 44) = 44
read(12, "\0\0\0\rf\0\0\0\1\0\0\0\4\0\0\0\0", 16384) = 17
write(4, "\222\251\32\377[\3412\235\334\363B\16E`te\tF\260}\303\267\220v\26V\375\203$\204\360\207"..., 80) = 80
```
80 byte read write system calls are inefficient. BAD.
* the move finishes after 22 minutes. BAD

Expected behaviour: Do 8192 byte chunks more often. loadavg ca 1.0 and a few seconds max. (A manual sftp "put" into that storage reaches > 100MBytes / sec)

**Comment by DeepDiver1975** [comment]
**Classification: confirmed-bug** (located cause) — fix deferred; this comment is analysis only.

Thanks for the detailed `strace` — it points straight at the cause, and this is a real, fixable inefficiency rather than an inherent SFTP/network limit (your manual `sftp put` reaching >100 MB/s on the same link proves the transport and crypto are not the bottleneck).

### Root cause

A web-UI move from primary storage to the SFTP mount copies the file through ownCloud's generic stream copy:

- `lib/private/legacy/helper.php` → `OC_Helper::streamCopy()` reads the source in **8 KB chunks** and `fwrite()`s each chunk to the destination stream.
- The SFTP destination stream is phpseclib's `phpseclib3\Net\SFTP\Stream` wrapper, opened in `apps/files_external/lib/Lib/Storage/SFTP.php::fopen()`.

The problem is in how that wrapper turns each `fwrite` into an SFTP operation. In phpseclib (`Net/SFTP/Stream.php::_stream_write()`):

```php
$result = $this->sftp->put($this->path, $data, SFTP::SOURCE_STRING, $this->pos);
```

`put()` is called **once per 8 KB block**, and each `put()` performs a *complete* `SSH_FXP_OPEN` → single `SSH_FXP_WRITE` → `SSH_FXP_CLOSE` cycle, each step blocking on a synchronous server response (`Net/SFTP/SFTP.php::put()` / `read_put_responses()` / `close_handle()`).

That is exactly what your strace shows per block:
- `write(11, ".../home/ftpdata/200MB"…, 44)` — the per-block `SSH_FXP_OPEN` (filename re-sent every time)
- `write(11, … , 8221)` — the single 8 KB `SSH_FXP_WRITE`
- `read(12, … "Success")` — the synchronous per-packet status wait (the round-trip stall)
- the 44/17/28-byte packets — open/close/status overhead, repeated for **every** 8 KB block

For 200 MB that is ~25,600 open/write/close cycles, each gated on network latency. With even a few ms RTT this trivially reaches your observed ~22 minutes (~150 KB/s) at near-zero CPU load (`loadavg 0.01`), because the process spends almost all its time blocked waiting for round-trips, not encrypting.

Crucially, phpseclib *already supports* pipelining: a single `put()` can keep up to `NET_SFTP_UPLOAD_QUEUE_SIZE` (1024) writes in flight with a 32 KB `max_sftp_packet`, and only then reads the batched responses — this is what a real `sftp put` does and why it saturates the link. The ownCloud stream wrapper defeats this by issuing one `put()` (hence one open/write/close) per 8 KB, so the pipelining loop never runs more than one iteration.

### Why it is a bug, not "SFTP is just slow"

- CPU is idle, so it is not the pure-PHP crypto cost.
- A manual `sftp put` to the same server is >600x faster.
- The slowness is structural: re-opening and closing the remote file for every 8 KB and waiting for a round-trip each time.

### Direction for a fix (for whoever picks this up)

The generic per-chunk stream-copy path is the wrong tool for SFTP writes. Options, roughly in order of impact:

1. **Bypass the stream wrapper for whole-file writes.** The backend already has `uploadFile()` which does a single `$this->getConnection()->put($remote, $localFile, NET_SFTP_LOCAL_FILE)` — that single call streams the local file with phpseclib's internal pipelining. Wiring SFTP's `writeStream()` / `moveFromStorage()` / `copyFromStorage()` to spool to a temp file and do one pipelined `put()` (instead of going through `OC_Helper::streamCopy` + the per-chunk stream wrapper) would let one `put()` handle the entire file.
2. If the stream wrapper must stay, it should at minimum hold one open SFTP handle for the lifetime of the stream and append, rather than re-`put()`-ing (re-open/close) on every block — but that is an upstream phpseclib stream-wrapper limitation.

### Data that would help confirm a fix

- Wall-clock time for the same 200 MB move before/after, plus the SSH RTT to the server (`ping`), to quantify the round-trip dependence.
- A quick comparison of `uploadFile()` (single pipelined `put`) vs. the stream path on the same file would directly demonstrate the gain.

🤖 Automated issue analysis by Claude Code review agent.

**Comment by dj4oC** [comment]
## Implementation prompt

**Problem:** Moving/copying a file from primary storage to an SFTP external-storage mount goes through the generic `OC_Helper::streamCopy()` (8 KB chunked `fwrite()`s) against phpseclib's SFTP stream wrapper. Each 8 KB `fwrite()` triggers a full `put()` call in phpseclib, i.e. a complete `SSH_FXP_OPEN` → `SSH_FXP_WRITE` → `SSH_FXP_CLOSE` round-trip per chunk, defeating phpseclib's built-in write pipelining and making large transfers latency-bound (observed: 200 MB taking ~22 minutes over a link that does >100 MB/s via a manual `sftp put`).

**Acceptance criteria:**
- Writing a large file to an SFTP external-storage mount (via move/copy through the web UI or WebDAV) completes in time proportional to link bandwidth, not to (file size / 8 KB) round-trips.
- No regression to correctness of partial/streamed writes that must go through the stream wrapper (e.g. chunked uploads still land correctly).

**Files:**
- `apps/files_external/lib/Lib/Storage/SFTP.php` — wire `writeStream()` / `moveFromStorage()` / `copyFromStorage()` to use the existing `uploadFile()` path (a single pipelined `$this->getConnection()->put($remote, $localFile, NET_SFTP_LOCAL_FILE)`) for whole-file transfers, spooling to a local temp file first instead of routing through the generic per-chunk stream copy.

**Constraints:**
- Test coverage ≥ 85% on touched code (e2e preferred — run the tool, do not trust claimed %)
- Commit signing: GPG signature + DCO Signed-off-by
- Rebase onto base branch before pushing
- No root-directory edits

🤖 Automated by Claude Code

---
_Generated by [Claude Code](https://claude.ai/code)_
